allennlp/models/biaffine_dependency_parser.py

```diff
@@ -0,0 +1,493 @@
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        words : Dict[str, torch.LongTensor], required
+            The output of ``TextField.as_array()``, which should typically be passed directly to a
+            ``TextFieldEmbedder``. This output is a dictionary mapping keys to ``TokenIndexer``
+            tensors.  At its most basic, using a ``SingleIdTokenIndexer`` this is: ``{"tokens":
+            Tensor(batch_size, num_tokens)}``. This dictionary will have the same keys as were used
+            for the ``TokenIndexers`` when you created the ``TextField`` representing your
+            sequence.  The dictionary is designed to be passed directly to a ``TextFieldEmbedder``,
+            which knows how to combine different word representations into a single vector per
+            token in your input.
+        pos_tags : ``torch.LongTensor``, optional (default = None)
+            The output of a ``SequenceLabelField`` containing POS tags.
+        head_tags : torch.LongTensor, optional (default = None)
+            A torch tensor representing the sequence of integer gold class labels for the arcs
+            in the dependency parse. Has shape ``(batch_size, num_tokens)``.
+        head_indices : torch.LongTensor, optional (default = None)
+            A torch tensor representing the sequence of integer indices denoting the parent of every
+            word in the dependency parse. Has shape ``(batch_size, num_tokens)``.
+
+        Returns
+        -------
+        An output dictionary consisting of:
+        loss : ``torch.FloatTensor``, optional
+            A scalar loss to be optimised.
+        arc_loss : ``torch.FloatTensor``
+            The loss contribution from the unlabeled arcs.
+        loss : ``torch.FloatTensor``, optional
+            The loss contribution from predicting the dependency
+            tags for the gold arcs.
+        heads : ``torch.FloatTensor``
+            The predicted head indices for each word. A tensor
+            of shape (batch_size, timesteps).
+        head_types : ``torch.FloatTensor``
+            The predicted head types for each arc. A tensor
+            of shape (batch_size, timesteps).
+        mask : ``torch.LongTensor``
+            A mask denoting the padded elements in the batch.
+        """
+        embedded_text_input = self.text_field_embedder(words)
+        if pos_tags is not None and self._pos_tag_embedding is not None:
+            embedded_pos_tags = self._pos_tag_embedding(pos_tags)
+            embedded_text_input = torch.cat([embedded_text_input, embedded_pos_tags], -1)
+        elif self._pos_tag_embedding is not None:
+            raise ConfigurationError("Model uses a POS embedding, but no POS tags were passed.")
+
+        mask = get_text_field_mask(words)
+        float_mask = mask.float()
+        encoded_text = self.encoder(embedded_text_input, mask)
+
+        # shape (batch_size, timesteps, arc_representation_dim)
```

Comment:
I still have a strong preference for something with more semantic content than "timesteps" here; if you really prefer "timesteps", then leave it, I just think it's a pretty vacuous (and misleading) term.  You have "num_tokens" in part of your docstring above, so whatever you pick, make sure you update the docstring to match.